modules/search/class.jetpack-search.php

```diff
@@ -0,0 +1,1289 @@
+		if ( ! isset( self::$instance ) ) {
+			self::$instance = new Jetpack_Search();
+
+			self::$instance->setup();
+		}
+
+		return self::$instance;
+	}
+
+	/**
+	 * Perform various setup tasks for the class
+	 *
+	 * Checks various pre-requisites and adds hooks
+	 *
+	 * @module search
+	 */
+	public function setup() {
+		if ( ! Jetpack::is_active() ) {
+			return;
+		}
+
+		$this->jetpack_blog_id = Jetpack::get_option( 'id' );
+
+		if ( ! $this->jetpack_blog_id ) {
+			return;
+		}
+
+		$this->init_hooks();
+	}
+
+	/**
+	 * Setup the various hooks needed for the plugin to take over Search duties
+	 *
+	 * @module search
+	 */
+	public function init_hooks() {
+		add_action( 'widgets_init', array( $this, 'action__widgets_init' ) );
+
+		if ( ! is_admin() ) {
+			add_filter( 'posts_pre_query', array( $this, 'filter__posts_pre_query' ), 10, 2 );
+
+			add_filter( 'jetpack_search_es_wp_query_args', array( $this, 'filter__add_date_filter_to_query' ),  10, 2 );
+		}
+	}
+
+	/*
+	 * Run a search on the WP.com public API.
+	 *
+	 * @module search
+	 *
+	 * @param array $es_args Args conforming to the WP.com /sites/<blog_id>/search endpoint
+	 *
+	 * @return object|WP_Error The response from the public api, or a WP_Error
+	 */
+	public function search( array $es_args ) {
+		$service_url = 'https://public-api.wordpress.com/rest/v1/sites/' . $this->jetpack_blog_id . '/search';
+
+		$start_time = microtime( true );
+
+		$request = wp_remote_post( $service_url, array(
+			'headers' => array(
+				'Content-Type' => 'application/json',
+			),
+			'timeout'    => 10,
+			'user-agent' => 'jetpack_search',
+			'body'       => json_encode( $es_args ),
+		) );
+
+		$end_time = microtime( true );
+
+		if ( is_wp_error( $request ) ) {
+			return $request;
+		}
+
+		$response = json_decode( wp_remote_retrieve_body( $request ), true );
+
+		$took = is_array( $response ) && $response['took'] ? $response['took'] : null;
+
+		$query = array(
+			'args'          => $es_args,
+			'response'      => $response,
+			'response_code' => wp_remote_retrieve_response_code( $request ),
+			'elapsed_time'   => ( $end_time - $start_time ) * 1000, // Convert from float seconds to ms
+			'es_time'       => $took,
+			'url'           => $service_url,
+		);
+
+		do_action( 'did_jetpack_search_query', $query );
+
+		return $response;
+	}
+
+	/**
+	 * Bypass the normal Search query and offload it to Jetpack servers
+	 *
+	 * This is the main hook of the plugin and is responsible for returning the posts that match the search query
+	 *
+	 * @module search
+	 *
+	 * @param array $posts Current array of posts (still pre-query)
+	 * @param WP_Query $query The WP_Query being filtered
+	 *
+	 * @return array Array of matching posts
+	 */
+	public function filter__posts_pre_query( $posts, $query ) {
+		if ( ! $query->is_main_query() || ! $query->is_search() ) {
+			return $posts;
+		}
+
+		$this->do_search( $query );
+
+		if ( ! is_array( $this->search_result ) ) {
```

Comment:
If I'm reading through this right, if the API fails for any reason -- no index, WP.com down, etc, this would have the site return traditional search results, correct?